### PR TITLE
Add support for mitmproxy 4.0.4

### PR DESCRIPTION
**Pull request by bebef1987**

### CODE REVIEW

**Comment by bebef1987** [comment]
Used the alternate-server-replay-4.0.4-no-best-match

So this is the default 4.0.4 replay addon with some small tweaks

File: mitmproxy.py [hunk 1]
```diff
@@ -19,33 +21,56 @@ def binary(self):
...
+                            self.path,
+                        ]
+                    ),
+                ]
+        elif self.mitmproxy_version == "4.0.4":
+            if self.record:
+                command = [
+                    self.binary,
+                    # "--scripts",
+                    # os.path.join(self.scripts, "inject-deterministic.py"),
```
Comment (davehunt) [inline mitmproxy.py]: Why no deterministic in 4.0.4? I presume you need version specific versions of the script?

File: mitmproxy.py [hunk 2]
```diff
@@ -19,33 +21,56 @@ def binary(self):
...
+                    "--save-stream-file",
+                    self.path
+                ]
+            else:
+                command = [
+                    self.binary,
+                    "--scripts",
+                    os.path.join(self.scripts, "alternate-server-replay-4.0.4-no-best-match.py"),
```
Comment (davehunt) [inline mitmproxy.py]: So this script supports the 404 response for requests outside of the recording, but doesn't do the best matching? How confident are we that we don't need the best match script? Can we point to a specific mitmproxy enhancement that covers this feature?

File: mitmproxy.py [hunk 3]
```diff
@@ -19,33 +21,56 @@ def binary(self):
...
-                    [
-                        os.path.join(self.scripts, "alternate-server-replay.py"),
-                        self.path,
-                    ]
-                ),
-            ]
-        self.process = subprocess.Popen(command)
+            raise Exception("Could not find the mitmproxy version")
+
+        self.process = subprocess.Popen(command, env=os.environ)
```
Comment (davehunt) [inline mitmproxy.py]: Isn't the env inherited from the existing environment by default? Why is this change needed?

**Review by davehunt — CHANGES_REQUESTED** [review]
I'm tempted to leave this as a branch we can use for testing, and hold off until we're ready to support mitmproxy 4.0.4 for all new recordings. That way we can just update the version instead of supporting multiple versions.

File: mitmproxy.py (hunk 1 above)
Comment (bebef1987) [reply mitmproxy.py]: We will need to update the addon to work with 4.0.4 
If we have the base ready it should not be a lot of work

File: mitmproxy.py (hunk 3 above)
Comment (bebef1987) [reply mitmproxy.py]: you are right I will remove that

File: mitmproxy.py (hunk 2 above)
Comment (bebef1987) [reply mitmproxy.py]: I'm still testing this. 
We can add both and see which one is the best

**Comment by davehunt** [comment]
Here's my work in progress branch for Python 3 and mitmproxy 4.0.4: https://github.com/davehunt/raptor-studio/tree/py3

**Comment by whimboo** [comment]
Does PR #13 super-seed this PR?

**Comment by bebef1987** [comment]
@davehunt  I updated this pull request to add mitmproxy 4.0.4 
Please review

**Comment by alexandru-io** [comment]
I tested (android/firefox/chrome recprd/replay) it and I am getting this. Otherwise the log looks fine.
```
Loading script /home/alexandru.ionescu/workspace/raptor-studio/scripts/alternate-server-replay-4.0.4.py
load options
Could not load command replay.server.file: missing type annotation
Could not load command replay.server: missing type annotation
```

**Comment by bebef1987** [comment]
That's from `alternate-server-replay-4.0.4.py` and that file is the same as the one in raptor

We get a error because it's missing py3 types. but the functionality is fine

File: scripts/alternate-server-replay-4.0.4.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
+# This file was copied from  mitmproxy/mitmproxy/addons/serverplayback.py release tag 4.0.4
```
Comment (davehunt) [inline scripts/alternate-server-replay-4.0.4.py:1]: As we're only supporting one version of mitmproxy, could we rename this file to remove the version number?

**Review by davehunt — CHANGES_REQUESTED** [review]
